op-bindings/bindgen/generator_local.go

```diff
@@ -0,0 +1,244 @@
+package main
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+	"path"
+	"path/filepath"
+	"regexp"
+	"strings"
+	"text/template"
+
+	"github.com/ethereum-optimism/optimism/op-bindings/ast"
+	"github.com/ethereum-optimism/optimism/op-bindings/foundry"
+)
+
+type bindGenGeneratorLocal struct {
+	bindGenGeneratorBase
+	sourceMapsList     string
+	forgeArtifactsPath string
+}
+
+type localContractMetadata struct {
+	Name              string
+	StorageLayout     string
+	DeployedBin       string
+	Package           string
+	DeployedSourceMap string
+}
+
+func (generator *bindGenGeneratorLocal) readContractsList() ([]string, error) {
+	var list contractsList
+	err := readJSONFile(generator.logger, generator.contractsListPath, &list)
+	return list.Local, err
+}
+
+func (generator *bindGenGeneratorLocal) generateBindings() error {
+	contracts, err := generator.readContractsList()
+	if err != nil {
+		return fmt.Errorf("error reading contract list %s: %w", generator.contractsListPath, err)
+	}
+	if len(contracts) == 0 {
+		return fmt.Errorf("no contracts parsed from given contract list: %s", generator.contractsListPath)
+	}
+
+	return generator.processContracts(contracts)
+}
+
+func (generator *bindGenGeneratorLocal) processContracts(contracts []string) error {
+	tempArtifactsDir, err := mkTempArtifactsDir(generator.logger)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		err := os.RemoveAll(tempArtifactsDir)
+		if err != nil {
+			generator.logger.Error("Error removing temporary artifact directory", "path", tempArtifactsDir, "err", err.Error())
+		} else {
+			generator.logger.Debug("Successfully removed temporary artifact directory")
+		}
+	}()
+
+	sourceMapsList := strings.Split(generator.sourceMapsList, ",")
+	sourceMapsSet := make(map[string]struct{})
+	for _, k := range sourceMapsList {
+		sourceMapsSet[k] = struct{}{}
+	}
+
+	contractArtifactPaths, err := generator.getContractArtifactPaths()
+	if err != nil {
+		return err
+	}
+
+	contractMetadataFileTemplate := template.Must(template.New("localContractMetadata").Parse(localContractMetadataTemplate))
+
+	for _, contractName := range contracts {
+		generator.logger.Info("Generating bindings and metadata for local contract", "contract", contractName)
+
+		forgeArtifact, err := generator.readForgeArtifact(contractName, contractArtifactPaths)
+		if err != nil {
+			return err
+		}
+
+		abiFilePath, bytecodeFilePath, err := writeContractArtifacts(generator.logger, tempArtifactsDir, contractName, forgeArtifact.Abi, []byte(forgeArtifact.Bytecode.Object.String()))
+		if err != nil {
+			return err
+		}
+
+		err = genContractBindings(generator.logger, abiFilePath, bytecodeFilePath, generator.bindingsPackageName, contractName)
```

Comment:
is it expected for this command to by run in the directory where the go file should be outputted?

wondering if it makes sense for there to be a flag for binding outputs?

Reply:
It's made to generate in `op-bindings/bindings` and expected to be there by some packages (like the L2 genesis generation scripts), so you couldn't change the output dir without changing imports paths - I think this is something we could refactor at a later time if needed